File: backend/src/main/java/com/imedba/modules/payment/entity/Payment.java
package com.imedba.modules.payment.entity;

import com.imedba.common.enums.PaymentMethod;
import com.imedba.modules.enrollment.entity.Enrollment;
import com.imedba.modules.installment.entity.Installment;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.Objects;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.UuidGenerator;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

/**
 * Registro de un pago. Es append-only: nunca se edita ni se elimina físicamente.
 * Para revertir un pago se crea un ajuste negativo (out of scope fase 2).
 */
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "payments")
@EntityListeners(AuditingEntityListener.class)
public class Payment {

    @Id
    @UuidGenerator
    @Column(name = "id", nullable = false, updatable = false)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "installment_id")
    private Installment installment;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "enrollment_id", nullable = false, updatable = false)
    private Enrollment enrollment;

    @Column(name = "amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal amount;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_method", nullable = false, length = 30)
    private PaymentMethod paymentMethod;

    @Column(name = "payment_date", nullable = false)
    private Instant paymentDate;

    @Column(name = "reference_number", length = 200)
    private String referenceNumber;

    @Column(name = "receipt_number", length = 50, unique = true)
    private String receiptNumber;

    @Column(name = "receipt_file_path", length = 500)
    private String receiptFilePath;

    @Column(name = "receipt_sent_at")
    private Instant receiptSentAt;

    @Column(name = "notes", columnDefinition = "text")
    private String notes;

    @Column(name = "registered_by")
    private UUID registeredBy;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Payment that)) return false;
        return id != null && Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: frontend/src/types/payment.ts
import type { Instant, UUID } from './common'
import type { PaymentMethod } from './enrollment'

export { PAYMENT_METHODS, PAYMENT_METHOD_LABELS } from './enrollment'
export type { PaymentMethod } from './enrollment'

// Refleja com.imedba.modules.payment.dto.PaymentResponse
// Plano: solo installmentId / enrollmentId. Alumno/curso se resuelven aparte.
export interface Payment {
  id:               UUID
  installmentId:    UUID | null      // null si es pago suelto (matrícula, libro, ajuste)
  enrollmentId:     UUID
  amount:           number
  paymentMethod:    PaymentMethod
  paymentDate:      Instant          // Instant (timestamp), no LocalDate
  referenceNumber:  string | null
  receiptNumber:    string | null    // IMD-... (server-side)
  receiptFilePath:  string | null
  receiptSentAt:    Instant | null
  notes:            string | null
  registeredBy:     UUID | null
  createdAt:        Instant
}

// Refleja com.imedba.modules.payment.dto.PaymentCreateRequest
export interface PaymentCreateRequest {
  installmentId?:   UUID | null      // obligatorio para pago imputado a cuota
  enrollmentId?:    UUID | null      // obligatorio si NO hay installmentId
  amount:           number           // > 0.01
  paymentMethod:    PaymentMethod
  paymentDate?:     Instant | null
  referenceNumber?: string | null
  receiptFilePath?: string | null
  notes?:           string | null
}
